log/ratelimit.go

```diff
@@ -0,0 +1,116 @@
+package log
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"golang.org/x/time/rate"
+)
+
+const (
+	infoLevel  = "info"
+	debugLevel = "debug"
+	warnLevel  = "warning"
+	errorLevel = "error"
+)
+
+type RateLimitedLogger struct {
+	next    Interface
+	limiter *rate.Limiter
+
+	discardedLogLinesCounter *prometheus.CounterVec
+}
+
+// NewRateLimitedLogger returns a logger.Interface that is limited to the given number of logs per second,
+// with the given burst size.
+func NewRateLimitedLogger(logger Interface, logsPerSecond rate.Limit, burstSize int, reg prometheus.Registerer) Interface {
+	discardedLogLinesCounter := promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+		Name: "rate_limit_logger_discarded_log_lines_total",
+		Help: "Total number of discarded log lines per level.",
+	}, []string{"level"})
+
+	return &RateLimitedLogger{
+		next:                     logger,
+		limiter:                  rate.NewLimiter(logsPerSecond, burstSize),
+		discardedLogLinesCounter: discardedLogLinesCounter,
+	}
+}
+
+func (l *RateLimitedLogger) Debugf(format string, args ...interface{}) {
+	if l.limiter.Allow() {
+		l.next.Debugf(format, args...)
+	} else {
+		l.discardedLogLinesCounter.WithLabelValues(debugLevel).Inc()
```

Comment:
Calling `l.discardedLogLinesCounter.WithLabelValues(debugLevel)` is very expensive. Instead, since the levels are known in advance, grab the metric instance in `NewRateLimitedLogger()` and store it in the `RateLimitedLogger` struct.

_Same comment applies below._